### ISSUE TITLE
Better support when auto-formatters have major changes

### ISSUE THREAD

**Issue by pbrisbin** [restyled-io/restylers#550]
From #549,

> In general since major clang-format releases break config-files and formatting, it might be benefiticial to support different versions of clang-format at the same time. I know it's possible now in different ways (pinning the restylers version or pinning the restyler's image in its configuration) but these are quite tedious since to look up the correct version or image coordinates. Ideally I'd like to enable clang-format-16 in my restyler config and switch to 17 once it is released and the code is in a stage where I can afford a full reformat

I see a few options for supporting this, and will list more as I think about this problem.

### 1- Separate Restylers

We (or anyone via PR) could simply `cp -r ./clang-format ./clang-format-16`  to define a completely separate `clang-format-16` restyler. As `clang-format` sees updates (e.g. to 17), we would just leave `clang-format-16` alone. We would retain `clang-format` as enabled by default, and `clang-format-16` not.

Users interested in such a feature would have to specify:

```yaml
restylers:
  - clang-format:
      enabled: false
  - clang-format-16
  - "*"
```

There is a question of when `clang-format-16` gets created. You can either do it the first time a v16 exists, or only when `clang-format` moves onto v17. The former is good for users because they can start using `clang-format-16` now and know it'll keep working whenever v17 lands, but it's bad for us because we'll have to make any updates in two places (or build some way to DRY it). The latter is better for us because we're effectively just freezing a v16 before we move onto v17, which we maintain normally, but it's worse for users because they will have to watch for and/or experience the v17 update before they could have a `clang-format-16` to use to avoid that breakage.

Another downside is that this is a very case-by-case approach. We'd have to think about it, do it, and maintain it by individual Restyler. This basically means it won't be available to users unless they happen to use a restyler where we do it.

### 2- "Series" Images

It would be relatively low effort to update the Restyler release process such that for any `restyled/restyler-{name}:vX.Y.Z` image we're releasing, we also push images tagged `:vX.Y` and `:vX`.

Users interested in such a feature would have to specify:

```yaml
restylers:
  - clang-format:
      image: restyled/restyler-clang-format:v16
  - "*"
```

This solves pretty much all the problems of (1) making it, IMO, strictly better.

- The required user configuration is smaller and perhaps less surprising
- These images will be available the first time any major version is seen, without incurring any extra maintenance
- We can do this for all Restylers all the time

The only potential downside is that mutable tags can be non-deterministic on use, but I think the risks are acceptable.

### 3- Multi-binary Images

We could build a `clang-format` image that contains any `clang-format-X` binaries we want. Whatever is latest would still remain the default, but users could specify:

```yaml
restylers:
  - clang-format:
      command:
        - clang-format-16
        - -i
```

The main downsides here are:

- Whenever overriding `command` users need to know about (e.g) the `-i` option
- We might run into cases where multiple versions can't co-exist
- This is (like Option 1) a case-by-case feature

**Comment by wirew0rm** [comment]
Wow, thanks for picking this up and documenting everything so nicely.

1) was what I was also coming up with, but having to manually disable the default one doesn't look so nice and I also see the maintainance issues.

2) This one sounds like the best one for me.
Would it be possible to make the image version a separate config entry? That way users wouldn't have to care about the actual image location and only for the version. That would probably be in the restyler repo and could be implemented afterwards.
Also if I understand correctly, this would only update the latest version while the older revisions would not get further updates. For me this would be acceptable but just noting it as a difference to the other approaches. If we still want that we would need multiple (or `ARG version=default` parameterized) Dockerfiles per restyler, plus some additional mechanics in the generation of the build matrix.
I've looked a bit into where the tag magic happens, if this was github actions only I could look into adding this, but if I understand it correctly the version logic at least partially happens in restyler's sdk github action which is written in haskell which I couldn't yet manage to get at least a basic understanding of ;)

3) The first downside could be prevented by adding a pre-command setting allowing to set up an alias/symlink for the version we want to use. Still prefer version 2, though ;)

**Comment by pbrisbin** [comment]
Yeah, it sounds 2 would be best. I agree.

> Would it be possible to make the image version a separate config entry? 

I had the same thought and yeah, we should tackle that as its own issue. Right now, I'm thinking of expanding the `image` schema so all of the following would be valid:

```yaml
image: "registry/image:tag" # today

image:
  registry: ghcr.io # => ghcr.io/restyler-foo:vX.Y.Z

image:
  name: something-else # => restyled/something-else:vX.Y.Z

image:
  tag: vA.B.Z # => restyled/restyler-foo:vA.B.Z

image:
  registry: ghcr.io
  tag: vA
  # => ghcr.io/restyler-foo:vA
```

> Also if I understand correctly, this would only update the latest version while the older revisions would not get further updates

I think that's right, but just to be totally clear:

1. We would only ever work in `clang-format/Dockerfile`, updating it just like today
2. When we build and release an image for any `vX.Y.Z`, we also tag it as `vX` and `vX.Y`

So: when we release `v16.0.0`, we would also tag it as `v16`. When we release `v16.0.1` we tag _it_ as `v16` (effectively updating the image to latest version in the series). When we release `v17.0.0`, we would tag that `v17`. At that point, nothing would ever cause the `v16` tag to change again. The idea is to work like conventional GitHub Actions versioning.

Does this make sense? Does it work?

It's possible that, after v17 has been out for a while, v16 gets some kind of security or maintenance release or something. In this case, nothing automatic would happen (nothing happens today) for our v16 image(s); it would just be a case by case if we want to manually get that fix out somehow. Usually, security patches aren't particularly important because we run our Restylers in such a restricted environment, so it's not a big deal to skip the effort.

> I've looked a bit into where the tag magic happens, if this was github actions only I could look into adding this

Yeah, it would actually be a lot easier and more reasonable if this were implemented in GHA with git tags, the `docker-metadata-action` and the `buildx` action, with its `tags` and `push` options. But for various reasons (some technical, some historical), the logic indeed lives in `restyled/sdk`.

**Comment by wirew0rm** [comment]
> So: when we release `v16.0.0`, we would also tag it as `v16`. When we release `v16.0.1` we tag _it_ as `v16` (effectively updating the image to latest version in the series). When we release `v17.0.0`, we would tag that `v17`. At that point, nothing would ever cause the `v16` tag to change again. The idea is to work like conventional GitHub Actions versioning.
> 
> Does this make sense? Does it work?

Yes, and I agree it's should be fine for restyler's usecase for the reasons you outlined, i just pointed it out because its different behaviour than 1) and 3) would exhibit.

> > I've looked a bit into where the tag magic happens, if this was github actions only I could look into adding this
> 
> Yeah, it would actually be a lot easier and more reasonable if this were implemented in GHA with git tags, the `docker-metadata-action` and the `buildx` action, with its `tags` and `push` options. But for various reasons (some technical, some historical), the logic indeed lives in `restyled/sdk`.

I'm not complaining, it seems to work well. Just wanted to make sure that I understood that correctly, makes it a bit harder for me to contribute there, but that's on me.

The proposed image scheme looks perfect :100: 

For now I'm using
```yaml
restylers:
  - clang-format:
      image: restyled/restyler-clang-format:16.0.6
  - "*"
```
so i have to manually add updated versions and its not super concise, but these are all more convenience improvements that I can change later.

**Comment by pbrisbin** [comment]
Alright, some stuff in progress:

1. `restyled/sdk` is updated to infer and release series tags: https://github.com/restyled-io/sdk/commit/947ec815645235eaeb80f8b3772077fca79f2272
2. `restyled/restyler` supports overriding image fields: https://github.com/restyled-io/restyler/pull/211

TODO:

1. Merge/test the above
2. Release `clang-format` as `v16...` instead of `16...` (the `v` prefix is required to get this behavior)
3. Document the new `image` config and series image idea

**Comment by pbrisbin** [comment]
OK, everything looks merged and deploy. I _think_ you can try using:

```yaml
restylers:
  - clang-format:
      image: {tag: v16}
  - "*"
```

Please let me know if it works!

**Comment by wirew0rm** [comment]
Works like a charm. Sorry for the late feedback, started this in the week before vacation and didn't want to do to many experiments before taking off.
